### PR TITLE
Ir core compiler

### PR DESCRIPTION
**Pull request by shirly121**
## Related issue number

Fixes

### CODE REVIEW

File: research/gaia-x/ir/compiler/conf/ir.plan.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pegasus.worker.num: 1
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/conf/ir.plan.properties:1]: These are engine.properties, No?

File: research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/PlanFactory.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        CPointer ptrPairs = irCoreJNI.initKVEqPairs();
+        irCoreJNI.addKVEqPair(ptrPairs, irCoreJNI.asIdKey(), irCoreJNI.int64AsConst(personId));
+        irCoreJNI.addIdxScanKVEqPairs(ptrIdxScan, ptrPairs);
+        irCoreJNI.appendIdxScanOp(ptrPlan, ptrIdxScan, (int) oprIdx.getValue(), oprIdx);
+
+        CPointer ptrOut = irCoreJNI.initExpandBase(Direction.Out);
+        irCoreJNI.addExpandLabel(ptrOut, irCoreJNI.cStrAsNameOrId("knows"));
+        CPointer ptrOutXpd = irCoreJNI.initEdgeXpdOp(ptrOut, true);
+        irCoreJNI.setEdgexpdAlias(ptrOutXpd, NameOrId.newBuilder().setOpt(NameIdOpt.Name).setName("p").build());
+        irCoreJNI.appendEdgeXpdOp(ptrPlan, ptrOutXpd, (int) oprIdx.getValue(), oprIdx);
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/PlanFactory.java:53]: All these functions return error code. Please handle them with proper Exception.

File: research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/PlanFactory.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        CPointer ptrScan = irCoreJNI.initScanOp(ScanOpt.Vertex);
+        CPointer ptrIdxScan = irCoreJNI.initIdxScanOp(ptrScan);
+        CPointer ptrPairs = irCoreJNI.initKVEqPairs();
+        irCoreJNI.addKVEqPair(ptrPairs, irCoreJNI.asIdKey(), irCoreJNI.int64AsConst(personId));
+        irCoreJNI.addIdxScanKVEqPairs(ptrIdxScan, ptrPairs);
+        irCoreJNI.appendIdxScanOp(ptrPlan, ptrIdxScan, (int) oprIdx.getValue(), oprIdx);
+
+        CPointer ptrOut = irCoreJNI.initExpandBase(Direction.Out);
+        irCoreJNI.addExpandLabel(ptrOut, irCoreJNI.cStrAsNameOrId("knows"));
+        CPointer ptrOutXpd = irCoreJNI.initEdgeXpdOp(ptrOut, true);
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/PlanFactory.java:51]: should be `irCoreJNI.initEdgeXpdOp(ptrOut, false);` ? As `true` means this is `outE`.  Same applies as follows.

File: research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/PlanFactory.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                        NameOrId.newBuilder().setOpt(NameIdOpt.Name).setName("HEAD").build(),
+                        new Property(
+                                PropertyOpt.Key,
+                                NameOrId.newBuilder().setOpt(NameIdOpt.Name).setName("id").build())),
+                OrderOpt.Asc);
+        irCoreJNI.setOrderByLimit(ptrOrderBy, 20, 21);
+        irCoreJNI.appendOrderByOp(ptrPlan, ptrOrderBy, (int) oprIdx.getValue(), oprIdx);
+
+        CPointer ptrProject = irCoreJNI.initProjectOp(false);
+        irCoreJNI.addProjectMapping(ptrProject,
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/PlanFactory.java:84]: I think you should have a `GetV` operator after "p" and "m" to obtain these properties.

File: research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/ResultCode.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package com.alibaba.graphscope.common.ir;
+
+public enum ResultCode {
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/java/src/main/java/com/alibaba/graphscope/common/ir/ResultCode.java:3]: We may cast these `ResultCode` into JavaException to better handle exception

**Review by longbinlai — COMMENTED** [review]
The most critical comment is that: we should handle the exception according to the ErrorCode that c gives you.

File: research/gaia-x/ir/compiler/conf/ir.plan.properties (hunk 1 above)
Comment (shirly121) [reply research/gaia-x/ir/compiler/conf/ir.plan.properties:1]: 是的

File: research/gaia-x/ir/compiler/src/main/java/com/alibaba/graphscope/common/SubmitPlanServiceMain.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            ptrPlan = PlanFactory.getCR2Plan();
+        } else {
+            throw new NotImplementedException("unimplemented opt type " + opt);
+        }
+
+        IrCoreLibrary irCoreLib = IrCoreLibrary.INSTANCE;
+        irCoreLib.debugPlan(ptrPlan);
+
+        FfiJobBuffer jobBuffer = irCoreLib.buildPhysicalPlan(ptrPlan);
+        assert jobBuffer.len > 0;
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/src/main/java/com/alibaba/graphscope/common/SubmitPlanServiceMain.java:40]: assertion is not a good idea for a service, log the error and return may be better.

File: research/gaia-x/ir/compiler/src/main/java/com/alibaba/graphscope/common/jna/IntEnum.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package com.alibaba.graphscope.common.jna;
+
+public interface IntEnum<T> {
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/src/main/java/com/alibaba/graphscope/common/jna/IntEnum.java:3]: What is this interface used for? `IntEnum` does not provide enough info.

File: research/gaia-x/ir/compiler/Makefile [hunk 8]
```diff
@@ -0,0 +1,11 @@
+OPT?=poc
+
+CUR_DIR:=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
+
+build:
+	cd $(CUR_DIR)/../core && cargo build --release && cp ./target/release/libir_core.dylib /usr/local/lib && cd $(CUR_DIR) && mvn clean package
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/Makefile:6]: I do not think it as a good idea to copy the library to `/usr/local/lib`, which may require root.

File: research/gaia-x/ir/compiler/src/main/java/com/alibaba/graphscope/common/PlanFactory.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+        IrCoreLibrary irCoreLib = IrCoreLibrary.INSTANCE;
+        IntByReference oprIdx = new IntByReference(0);
+        Pointer ptrPlan = irCoreLib.initLogicalPlan();
+
+        Pointer ptrScan = irCoreLib.initScanOperator(FfiScanOpt.Vertex);
+        Pointer ptrIdxScan = irCoreLib.initIdxscanOperator(ptrScan);
+
+        Pointer ptrPairs = irCoreLib.initKvEquivPairs();
+        checkResult(IrCoreLibrary.FunctionName.andKvEquivPair,
```
Comment (longbinlai) [inline research/gaia-x/ir/compiler/src/main/java/com/alibaba/graphscope/common/PlanFactory.java:61]: Can we `try` the whole program segment and then throw from there? It is better than adding a `checkResult` function which seems not very clean.

**Review by longbinlai — COMMENTED** [review]
Try to figure out if the dynamic library is not placed under /usr/local/lib

File: research/gaia-x/ir/compiler/conf/ir.plan.properties (hunk 1 above)
Comment (shirly121) [reply research/gaia-x/ir/compiler/conf/ir.plan.properties]: 我的想法是compiler所有配置都放在这个文件里，engine的只是作为其中一部分；ir.plan.properties这个名字可能也不太合理，我改成ir.compiler.properties，里面用注释区分不同配置的功能（java properties还不支持section）

Comment (longbinlai) [reply research/gaia-x/ir/compiler/conf/ir.plan.properties]: 嗯，这样比较好。

**Pull request merged**